mivideo/mivideo_extract.py

```diff
@@ -67,62 +85,246 @@ def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[date
-        }]
+        }
+
+    @staticmethod
+    def _queryRunner(
+            pandasTable: SQLTable,
+            dbConn: Connection,
+            columnNameList: Sequence[str],
+            data: Iterable):
+        '''
+        This handles the fairly rare occurrence of conflicting keys when
+        inserting data into a table.
+        '''
+
+        tableFullName = pandasTable.name
+        if (pandasTable.schema):
+            tableFullName = f'{pandasTable.schema}.{tableFullName}'
+        columnNames = ', '.join(columnNameList)
+        valuePlaceholders = ', '.join(['%s'] * len(columnNameList))
+
+        sql = (
+            f'INSERT INTO {tableFullName} ({columnNames}) VALUES ({valuePlaceholders}) '
+            f'ON DUPLICATE KEY UPDATE {columnNameList[0]}={columnNameList[0]}'  # magic
+        )
+
+        dbConn.execute(sql, list(data))
+
+    def mediaCreation(self) -> Dict[str, Union[ValidDataSourceName, pd.Timestamp]]:
+        """
+        Update data with Kaltura media metadata from Kaltura API.
+
+        :return: a dictionary with ValidDataSourceName and last run timestamp
+        """
+
+        self._kalturaInit()
+
+        KALTURA_MAX_MATCHES_ERROR: str = 'QUERY_EXCEEDED_MAX_MATCHES_ALLOWED'
+        tableName: str = 'mivideo_media_created'
+
+        localLogger = logging.getLogger(f'{logger.name}.mediaCreation')
+        localLogger.info('Starting procedure...')
+
+        kClient: KalturaRequestConfiguration = KalturaClient(KalturaConfiguration())
+        kClient.setKs(  # pylint: disable=no-member
+            KalturaSessionService(kClient).start(
+                self.kUserSecret, type=KalturaSessionType.ADMIN, partnerId=self.kPartnerId))
+        kMedia = KalturaMediaService(kClient)
+
+        lastTime: Union[datetime, None] = (
+            self._readTableLastTime(tableName, 'created_at'))
+
+        if (lastTime):
+            localLogger.info(
+                f'Last time found in table: "{lastTime.isoformat()}"')
+        else:
+            lastTime = datetime.fromisoformat(self.defaultLastTimestamp)
+            localLogger.info('Last time not found in table; '
+                             f'using default time: "{lastTime.isoformat()}"')
+
+        createdAtTimestamp: float = lastTime.timestamp()
+
+        kFilter = KalturaMediaEntryFilter()
+        kFilter.createdAtGreaterThanOrEqual = createdAtTimestamp
+        kFilter.categoriesFullNameIn = 'Canvas_UMich'
+        kFilter.orderBy = KalturaMediaEntryOrderBy.CREATED_AT_ASC
+
+        kPager = KalturaFilterPager()
+        kPager.pageSize = 500  # 500 is maximum
+        kPager.pageIndex = 1
+
+        results: Sequence[KalturaMediaEntry] = None
+        lastCreatedAtTimestamp: Union[float, int] = createdAtTimestamp
+        lastId: Union[str, None] = None
+        numberResults: int = 0
+        queryPageNumber: int = kPager.pageIndex  # for logging purposes
+        totalNumberResults: int = numberResults  # for logging purposes
+
+        while True:
```

Comment:
OK, so…

```python
trueFlag = True
while trueFlag:
  …
```

\<Ducks and runs\>

Reply:
Seriously, though.  I'll give it some thought.

Reply:
I'm thinking more like `more_pages = True`. Then when you set it to `False`, people actually know you intend to break a certain loop, rather than them wondering, "does this `break` apply to that `for` or this `while`? It's just clearer code in my opinion.

Reply:
Yeah, I like `endOfResults` than `while True`.